File: AmneziaAndroid/core/src/main/java/org/amnezia/core/parsers/WgQuickConfigParser.kt
package org.amnezia.core.parsers

import org.amnezia.core.models.ProtocolType
import org.amnezia.core.models.ServerProfile

object WgQuickConfigParser {

    fun parse(text: String, profileName: String? = null): ServerProfile {
        val interfaceAttrs = mutableMapOf<String, String>()
        val peerAttrs = mutableMapOf<String, String>()

        var currentSection = ""

        text.lines().forEach { line ->
            var trimmed = line.trim()
            // Strip comments
            val commentIdx = trimmed.indexOf('#')
            if (commentIdx != -1) {
                trimmed = trimmed.substring(0, commentIdx).trim()
            }
            if (trimmed.isEmpty()) return@forEach

            val lower = trimmed.lowercase()
            if (lower == "[interface]") {
                currentSection = "interface"
                return@forEach
            } else if (lower == "[peer]") {
                currentSection = "peer"
                return@forEach
            }

            val equalsIdx = trimmed.indexOf('=')
            if (equalsIdx != -1) {
                val key = trimmed.substring(0, equalsIdx).trim().lowercase()
                val value = trimmed.substring(equalsIdx + 1).trim()
                if (currentSection == "interface") {
                    val existing = interfaceAttrs[key]
                    interfaceAttrs[key] = if (existing != null) "$existing,$value" else value
                } else if (currentSection == "peer") {
                    val existing = peerAttrs[key]
                    peerAttrs[key] = if (existing != null) "$existing,$value" else value
                }
            }
        }

        if (interfaceAttrs.isEmpty() || peerAttrs.isEmpty()) {
            throw IllegalArgumentException("Missing [Interface] or [Peer] section in configuration")
        }

        val privateKey = interfaceAttrs["privatekey"]
            ?: throw IllegalArgumentException("Interface is missing PrivateKey")
        val publicKey = peerAttrs["publickey"]
            ?: throw IllegalArgumentException("Peer is missing PublicKey")
        val endpoint = peerAttrs["endpoint"]
            ?: throw IllegalArgumentException("Peer is missing Endpoint")

        val (host, port) = parseEndpoint(endpoint)

        val addresses = interfaceAttrs["address"]?.split(",")?.map { it.trim() }?.filter { it.isNotEmpty() }
            ?: listOf("10.0.0.2/32")
        val dnsList = interfaceAttrs["dns"]?.split(",")?.map { it.trim() }?.filter { it.isNotEmpty() }
            ?: listOf("1.1.1.1", "1.0.0.1")
        val allowedIps = peerAttrs["allowedips"]?.split(",")?.map { it.trim() }?.filter { it.isNotEmpty() }
            ?: listOf("0.0.0.0/0", "::/0")

        val presharedKey = peerAttrs["presharedkey"]
        val mtu = interfaceAttrs["mtu"]?.toIntOrNull() ?: 1280

        // AmneziaWG obfuscation parameters
        val jc = interfaceAttrs["jc"]?.toIntOrNull()
        val jmin = interfaceAttrs["jmin"]?.toIntOrNull()
        val jmax = interfaceAttrs["jmax"]?.toIntOrNull()
        val s1 = interfaceAttrs["s1"]?.toIntOrNull()
        val s2 = interfaceAttrs["s2"]?.toIntOrNull()
        val h1 = interfaceAttrs["h1"]?.toLongOrNull()
        val h2 = interfaceAttrs["h2"]?.toLongOrNull()
        val h3 = interfaceAttrs["h3"]?.toLongOrNull()
        val h4 = interfaceAttrs["h4"]?.toLongOrNull()

        val isAwg = jc != null || jmin != null || jmax != null || s1 != null || s2 != null ||
                h1 != null || h2 != null || h3 != null || h4 != null
        val protocolType = if (isAwg) ProtocolType.AMNEZIA_WG else ProtocolType.WIREGUARD
        val name = profileName ?: "WireGuard ($host)"

        return ServerProfile(
            name = name,
            protocolType = protocolType,
            host = host,
            port = port,
            clientIp = addresses.firstOrNull() ?: "10.0.0.2/32",
            dnsServers = dnsList,
            publicKey = publicKey,
            privateKey = privateKey,
            presharedKey = presharedKey,
            allowedIps = allowedIps,
            jc = jc,
            jmin = jmin,
            jmax = jmax,
            s1 = s1,
            s2 = s2,
            h1 = h1,
            h2 = h2,
            h3 = h3,
            h4 = h4,
            mtu = mtu
        )
    }

    private fun parseEndpoint(endpoint: String): Pair<String, Int> {
        val trimmed = endpoint.trim()
        if (trimmed.startsWith("[")) {
            // IPv6 [2001:db8::1]:51820
            val closing = trimmed.indexOf(']')
            if (closing != -1 && closing + 1 < trimmed.length && trimmed[closing + 1] == ':') {
                val host = trimmed.substring(1, closing)
                val port = trimmed.substring(closing + 2).toIntOrNull() ?: 51820
                return Pair(host, port)
            }
        }
        val colonIdx = trimmed.lastIndexOf(':')
        if (colonIdx != -1) {
            val host = trimmed.substring(0, colonIdx)
            val port = trimmed.substring(colonIdx + 1).toIntOrNull() ?: 51820
            return Pair(host, port)
        }
        return Pair(trimmed, 51820)
    }

    fun serialize(profile: ServerProfile): String {
        val sb = StringBuilder()
        sb.appendLine("[Interface]")
        sb.appendLine("PrivateKey = ${profile.privateKey}")
        sb.appendLine("Address = ${profile.clientIp}")
        if (profile.dnsServers.isNotEmpty()) {
            sb.appendLine("DNS = ${profile.dnsServers.joinToString(", ")}")
        }
        if (profile.mtu > 0) {
            sb.appendLine("MTU = ${profile.mtu}")
        }
        profile.jc?.let { sb.appendLine("Jc = $it") }
        profile.jmin?.let { sb.appendLine("Jmin = $it") }
        profile.jmax?.let { sb.appendLine("Jmax = $it") }
        profile.s1?.let { sb.appendLine("S1 = $it") }
        profile.s2?.let { sb.appendLine("S2 = $it") }
        profile.h1?.let { sb.appendLine("H1 = $it") }
        profile.h2?.let { sb.appendLine("H2 = $it") }
        profile.h3?.let { sb.appendLine("H3 = $it") }
        profile.h4?.let { sb.appendLine("H4 = $it") }

        sb.appendLine()
        sb.appendLine("[Peer]")
        sb.appendLine("PublicKey = ${profile.publicKey}")
        profile.presharedKey?.let { sb.appendLine("PresharedKey = $it") }
        sb.appendLine("Endpoint = ${profile.host}:${profile.port}")
        sb.appendLine("AllowedIPs = ${profile.allowedIps.joinToString(", ")}")
        sb.appendLine("PersistentKeepalive = 25")
        return sb.toString()
    }
}

File: AmneziaCore/Parsers/WgQuickConfigParser.swift
import Foundation

public enum WgQuickParserError: Error, LocalizedError {
    case missingInterfaceSection
    case missingPeerSection
    case missingPrivateKey
    case missingPublicKey
    case missingEndpoint
    case invalidEndpoint(String)
    case invalidPort(String)

    public var errorDescription: String? {
        switch self {
        case .missingInterfaceSection:
            return "Configuration is missing [Interface] section"
        case .missingPeerSection:
            return "Configuration is missing [Peer] section"
        case .missingPrivateKey:
            return "Interface is missing PrivateKey"
        case .missingPublicKey:
            return "Peer is missing PublicKey"
        case .missingEndpoint:
            return "Peer is missing Endpoint"
        case .invalidEndpoint(let ep):
            return "Invalid endpoint format: '\(ep)'"
        case .invalidPort(let p):
            return "Invalid port: '\(p)'"
        }
    }
}

public final class WgQuickConfigParser: Sendable {

    /// Parses a wg-quick / awg-quick configuration text into a `ServerProfile`
    public static func parse(_ text: String, profileName: String? = nil) throws -> ServerProfile {
        var interfaceAttrs = [String: String]()
        var peerAttrs = [String: String]()

        enum Section {
            case none
            case interface
            case peer
        }

        var currentSection = Section.none
        let lines = text.components(separatedBy: .newlines)

        for line in lines {
            var trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
            // Strip comments
            if let commentIndex = trimmed.firstIndex(of: "#") {
                trimmed = String(trimmed[..<commentIndex]).trimmingCharacters(in: .whitespacesAndNewlines)
            }
            guard !trimmed.isEmpty else { continue }

            let lower = trimmed.lowercased()
            if lower == "[interface]" {
                currentSection = .interface
                continue
            } else if lower == "[peer]" {
                currentSection = .peer
                continue
            }

            guard let equalsIndex = trimmed.firstIndex(of: "=") else { continue }
            let key = String(trimmed[..<equalsIndex]).trimmingCharacters(in: .whitespacesAndNewlines)
            let value = String(trimmed[trimmed.index(after: equalsIndex)...]).trimmingCharacters(in: .whitespacesAndNewlines)

            switch currentSection {
            case .interface:
                if let existing = interfaceAttrs[key.lowercased()] {
                    interfaceAttrs[key.lowercased()] = existing + "," + value
                } else {
                    interfaceAttrs[key.lowercased()] = value
                }
            case .peer:
                if let existing = peerAttrs[key.lowercased()] {
                    peerAttrs[key.lowercased()] = existing + "," + value
                } else {
                    peerAttrs[key.lowercased()] = value
                }
            case .none:
                break
            }
        }

        guard !interfaceAttrs.isEmpty else {
            throw WgQuickParserError.missingInterfaceSection
        }
        guard !peerAttrs.isEmpty else {
            throw WgQuickParserError.missingPeerSection
        }

        guard let privateKey = interfaceAttrs["privatekey"], !privateKey.isEmpty else {
            throw WgQuickParserError.missingPrivateKey
        }
        guard let publicKey = peerAttrs["publickey"], !publicKey.isEmpty else {
            throw WgQuickParserError.missingPublicKey
        }
        guard let endpoint = peerAttrs["endpoint"], !endpoint.isEmpty else {
            throw WgQuickParserError.missingEndpoint
        }

        // Parse Endpoint host:port
        let (host, port) = try parseEndpoint(endpoint)

        // Parse Addresses
        let addresses: [String]
        if let addr = interfaceAttrs["address"] {
            addresses = addr.components(separatedBy: ",").map { $0.trimmingCharacters(in: .whitespaces) }.filter { !$0.isEmpty }
        } else {
            addresses = ["10.8.0.2/32"]
        }

        // Parse DNS
        let dnsList: [String]
        if let dns = interfaceAttrs["dns"] {
            dnsList = dns.components(separatedBy: ",").map { $0.trimmingCharacters(in: .whitespaces) }.filter { !$0.isEmpty }
        } else {
            dnsList = ["1.1.1.1", "1.0.0.1"]
        }

        // Parse AllowedIPs
        let allowedIps: [String]
        if let ips = peerAttrs["allowedips"] {
            allowedIps = ips.components(separatedBy: ",").map { $0.trimmingCharacters(in: .whitespaces) }.filter { !$0.isEmpty }
        } else {
            allowedIps = ["0.0.0.0/0", "::/0"]
        }

        let presharedKey = peerAttrs["presharedkey"]
        let mtu = interfaceAttrs["mtu"].flatMap { UInt16($0) }
        let persistentKeepalive = peerAttrs["persistentkeepalive"].flatMap { UInt16($0) }

        // Parse AWG Obfuscation parameters
        let awg = AwgParameters(
            jc: interfaceAttrs["jc"].flatMap { UInt16($0) },
            jmin: interfaceAttrs["jmin"].flatMap { UInt16($0) },
            jmax: interfaceAttrs["jmax"].flatMap { UInt16($0) },
            s1: interfaceAttrs["s1"].flatMap { UInt16($0) },
            s2: interfaceAttrs["s2"].flatMap { UInt16($0) },
            s3: interfaceAttrs["s3"].flatMap { UInt16($0) },
            s4: interfaceAttrs["s4"].flatMap { UInt16($0) },
            h1: interfaceAttrs["h1"],
            h2: interfaceAttrs["h2"],
            h3: interfaceAttrs["h3"],
            h4: interfaceAttrs["h4"],
            i1: interfaceAttrs["i1"],
            i2: interfaceAttrs["i2"],
            i3: interfaceAttrs["i3"],
            i4: interfaceAttrs["i4"],
            i5: interfaceAttrs["i5"]
        )

        let protocolType: ProtocolType = awg.hasObfuscation ? .amneziaWg : .wireGuard
        let name = profileName ?? "AWG (\(host))"

        return ServerProfile(
            name: name,
            protocolType: protocolType,
            endpointHost: host,
            endpointPort: port,
            clientPrivateKey: privateKey,
            serverPublicKey: publicKey,
            presharedKey: presharedKey,
            clientAddresses: addresses,
            dnsServers: dnsList,
            allowedIps: allowedIps,
            mtu: mtu,
            persistentKeepalive: persistentKeepalive,
            awgParameters: awg.hasObfuscation ? awg : nil,
            rawConfig: text
        )
    }

    /// Serializes a `ServerProfile` into standard or AmneziaWG `.conf` text
    public static func serialize(_ profile: ServerProfile) -> String {
        var output = "[Interface]\n"
        output += "PrivateKey = \(profile.clientPrivateKey)\n"
        output += "Address = \(profile.clientAddresses.joined(separator: ", "))\n"
        if !profile.dnsServers.isEmpty {
            output += "DNS = \(profile.dnsServers.joined(separator: ", "))\n"
        }
        if let mtu = profile.mtu {
            output += "MTU = \(mtu)\n"
        }

        // AmneziaWG obfuscation parameters
        if let awg = profile.awgParameters {
            if let jc = awg.jc { output += "Jc = \(jc)\n" }
            if let jmin = awg.jmin { output += "Jmin = \(jmin)\n" }
            if let jmax = awg.jmax { output += "Jmax = \(jmax)\n" }
            if let s1 = awg.s1 { output += "S1 = \(s1)\n" }
            if let s2 = awg.s2 { output += "S2 = \(s2)\n" }
            if let s3 = awg.s3 { output += "S3 = \(s3)\n" }
            if let s4 = awg.s4 { output += "S4 = \(s4)\n" }
            if let h1 = awg.h1 { output += "H1 = \(h1)\n" }
            if let h2 = awg.h2 { output += "H2 = \(h2)\n" }
            if let h3 = awg.h3 { output += "H3 = \(h3)\n" }
            if let h4 = awg.h4 { output += "H4 = \(h4)\n" }
            if let i1 = awg.i1 { output += "I1 = \(i1)\n" }
            if let i2 = awg.i2 { output += "I2 = \(i2)\n" }
            if let i3 = awg.i3 { output += "I3 = \(i3)\n" }
            if let i4 = awg.i4 { output += "I4 = \(i4)\n" }
            if let i5 = awg.i5 { output += "I5 = \(i5)\n" }
        }

        output += "\n[Peer]\n"
        output += "PublicKey = \(profile.serverPublicKey)\n"
        if let psk = profile.presharedKey, !psk.isEmpty {
            output += "PresharedKey = \(psk)\n"
        }
        output += "Endpoint = \(profile.endpointHost):\(profile.endpointPort)\n"
        output += "AllowedIPs = \(profile.allowedIps.joined(separator: ", "))\n"
        if let keepalive = profile.persistentKeepalive {
            output += "PersistentKeepalive = \(keepalive)\n"
        }

        return output
    }

    // MARK: - Helpers

    private static func parseEndpoint(_ endpoint: String) throws -> (String, UInt16) {
        // Handle IPv6 bracket notation [2001:db8::1]:51820 or ipv4:port
        if endpoint.hasPrefix("[") {
            guard let closeBracket = endpoint.firstIndex(of: "]"),
                  let colonIndex = endpoint[closeBracket...].firstIndex(of: ":") else {
                throw WgQuickParserError.invalidEndpoint(endpoint)
            }
            let host = String(endpoint[endpoint.index(after: endpoint.startIndex)..<closeBracket])
            let portStr = String(endpoint[endpoint.index(after: colonIndex)...])
            guard let port = UInt16(portStr) else {
                throw WgQuickParserError.invalidPort(portStr)
            }
            return (host, port)
        } else {
            let parts = endpoint.components(separatedBy: ":")
            guard parts.count == 2 else {
                throw WgQuickParserError.invalidEndpoint(endpoint)
            }
            let host = parts[0]
            guard let port = UInt16(parts[1]) else {
                throw WgQuickParserError.invalidPort(parts[1])
            }
            return (host, port)
        }
    }
}
